### ISSUE TITLE
DeepSeek V4 sparse-attention mask drops pool position 0 due to `scatter_` overwrite on clamped `-1` indices

### ISSUE THREAD

**Issue by jQizhang** [NVIDIA-NeMo/Automodel#2144]
**Describe the bug**

In `DeepseekV4Attention.forward` (`nemo_automodel/components/models/deepseek_v4/layers.py`, around L905–L913), `indexer_topk` (shape `[B, S, K]`, with `-1` marking causally-masked / invalid slots) is converted into a boolean `indicator` mask of shape `[B, S, P]` via:

```python
valid = indexer_topk != -1                              # [B, S, K]
safe_idx = indexer_topk.clamp(min=0)                    # -1 -> 0 to avoid OOB
indicator = torch.zeros((B, S, P), dtype=torch.bool, device=...)
indicator.scatter_(-1, safe_idx, valid)
```

This is incorrect. `clamp(min=0)` collapses **all `-1` entries onto the same destination index `0`**, and `scatter_` uses overwrite semantics when there are duplicate indices along the scattered dim. As a result, whenever a query's `indexer_topk[b, s, :]` contains both:

1. a valid entry pointing to pool position `0` (`valid=True`), **and**
2. one or more `-1` entries (`valid=False`, but `safe_idx=0` after clamp),

the `False` writes can overwrite the `True` write at `indicator[b, s, 0]`, depending on the order along K. The mask then incorrectly forces `-inf` at pool position `0` for that query, so the query silently fails to attend to a pool slot it was supposed to attend to.

**Steps/Code to reproduce bug**

Standalone reproducer of just the mask-construction snippet:

```python
import torch

B, S, K, P = 1, 1, 4, 5

# Query wants to attend to pool positions 2 and 0;
# the trailing two slots are causally masked (-1).
indexer_topk = torch.tensor([[[2, 0, -1, -1]]])   # [B, S, K]

valid = indexer_topk != -1                        # [[[T, T, F, F]]]
safe_idx = indexer_topk.clamp(min=0)              # [[[2, 0, 0, 0]]]

# --- current (buggy) implementation ---
indicator = torch.zeros((B, S, P), dtype=torch.bool)
indicator.scatter_(-1, safe_idx, valid)
print("buggy   :", indicator)

# --- expected ---
print("expected:", torch.tensor([[[True, False, True, False, False]]]))

# --- proposed fix ---
counts = torch.zeros((B, S, P), dtype=torch.int32)
counts.scatter_add_(-1, safe_idx, valid.to(torch.int32))
fixed = counts > 0
print("fixed   :", fixed)
```

Output:

```
buggy   : tensor([[[False, False,  True, False, False]]])
expected: tensor([[[ True, False,  True, False, False]]])
fixed   : tensor([[[ True, False,  True, False, False]]])
```

**Additional context**

Suggested fix — replace overwrite with count-then-threshold, so invalid slots contribute `0` and cannot stomp on valid writes:

```python
valid = indexer_topk != -1
safe_idx = indexer_topk.clamp(min=0)
indicator_counts = torch.zeros(
    (B, S, P), dtype=torch.int32, device=full_kv.device,
)
indicator_counts.scatter_add_(-1, safe_idx, valid.to(torch.int32))
indicator = indicator_counts > 0
```

Downstream impact: in NeMo-RL GRPO training on DSV4, after applying this fix the observed `gen_kl_error` drops from **0.02591** to **0.02494**.

**Comment by HuiyingLi** [comment]
@jQizhang Thank you so much! Would you mind send a PR for the fix?

**Comment by jQizhang** [comment]
Hi @HuiyingLi  Sure, I’ll work on a fix and open a PR shortly.

**Comment by jQizhang** [comment]
hi @HuiyingLi the PR is ready for review at #2152. Let me know if any further changes are needed. Thanks!
